### ISSUE TITLE
HiddenField module does not account for static visibility from a closure

### ISSUE THREAD

**Issue by henryptung** [checkstyle/checkstyle#382]
Consider the following input file:
```java
class A {
    final A field;
    static void method(A field) {}
    static class B {
        void method(A field) {}
    }
    static Comparator<A> COMP = new Comparator<A>() {
        @Override
        public int compare(A field, A other) { //warn is here
            return 0;
        }
    };
    static Comparator<A> createComp() {
        return new Comparator<A>() {
            @Override
            public int compare(A field, A other) { //warn is here
                return 0;
            }
        };
    }
}
```
with the following configuration:
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<!-- http://checkstyle.sourceforge.net/ -->
<module name="Checker">
    <module name="TreeWalker">
        <module name="HiddenField"/>
    </module>
</module>
```

It would be expected in this case that no warnings are issued, because the field `A.field` is not visible within any of the methods/closures specified.  However, the output with Checkstyle 6.1 is:

```
Starting audit...
<path omitted>/A.java:9:30: 'field' hides a field.
<path omitted>/A.java:16:34: 'field' hides a field.
Audit done.
```

The two closure methods are erroneously marked as being within the context of an `A` instance, triggering the error.

**Comment by romani** [comment]
Assigned to Dmitrii Priimak.

**Comment by MEZk** [comment]
@romani 
Do we need some other improvements?

**Comment by romani** [comment]
@mkordas pointed to missed tests for lambdas in PR, please extend inputs .

**Comment by MEZk** [comment]
@romani 
>@mkordas pointed to missed tests for lambdas in PR, please extend inputs .

Done in scope of https://github.com/checkstyle/checkstyle/issues/2175
